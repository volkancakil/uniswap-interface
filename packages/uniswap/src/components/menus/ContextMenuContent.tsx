import { Fragment } from 'react'
import { Flex, FlexProps, Separator, Text, TouchableArea } from 'ui/src'
import { MenuOptionItem } from 'uniswap/src/components/menus/ContextMenuV2'

type MenuContentProps = {
  items: MenuOptionItem[]
  onItemClick?: () => void
} & FlexProps

export function MenuContent({ items, onItemClick, ...rest }: MenuContentProps): JSX.Element {
  return (
    <Flex
      asChild
      flexDirection="column"
      gap="$spacing4"
      p="$spacing8"
      backgroundColor="$surface1"
      borderRadius="$rounded20"
      borderWidth="$spacing1"
      borderColor="$surface3"
      {...rest}
    >
      <div
        // Prevent any right-click from bubbling up or showing default context menu
        onContextMenu={(e) => {
          e.preventDefault()
          e.stopPropagation()
        }}
      >
        {items.map(
          ({ label, onPress, Icon, showDivider, disabled, iconProps, closeDelay, ...touchableProps }, index) => (
            <Fragment key={index}>
              {showDivider && <Separator my="$spacing6" />}
              <TouchableArea
                hoverable
                disabled={disabled}
                borderRadius="$rounded12"
                width="100%"
                userSelect="none"
                cursor={disabled ? 'default' : 'pointer'}
                onPress={(e) => {
                  e.stopPropagation()
                  e.preventDefault()
                  onPress(e)

                  if (closeDelay) {
                    setTimeout(() => onItemClick?.(), closeDelay)
                  } else {
                    onItemClick?.()
                  }
                }}
                {...touchableProps}
              >
                <Flex row alignItems="center" p="$padding8" gap="$gap8" alignContent="center">
                  {Icon && <Icon size="$icon.16" color="$neutral2" {...iconProps} />}
                  <Text
                    variant="buttonLabel3"
                    color={disabled ? '$neutral2' : '$neutral1'}
                    hoverStyle={{ opacity: disabled ? 1 : 0.8 }}
                  >
                    {label}
                  </Text>
                </Flex>
              </TouchableArea>
            </Fragment>
          ),
        )}
      </div>
    </Flex>
  )
}
