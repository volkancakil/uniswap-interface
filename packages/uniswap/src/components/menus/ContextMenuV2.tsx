import { BaseSyntheticEvent, Fragment, PropsWithChildren, useRef, useState } from 'react'
import { FlexProps, GeneratedIcon, IconProps, Popover, PopperProps, TextProps, TouchableAreaProps } from 'ui/src'
import { MenuContent } from 'uniswap/src/components/menus/ContextMenuContent'
import { useOnClickOutside } from 'utilities/src/react/hooks'

export type MenuOptionItem = {
  label: string
  onPress: (e: BaseSyntheticEvent) => void
  textProps?: TextProps
  Icon?: GeneratedIcon | ((props: IconProps) => JSX.Element)
  showDivider?: boolean
  disabled?: boolean
  iconProps?: IconProps
  closeDelay?: number
} & TouchableAreaProps

type ContextMenuProps = {
  menuItems: MenuOptionItem[]
  menuStyleProps?: FlexProps
  onLeftClick?: boolean
  alignContentLeft?: boolean
  disabled?: boolean
} & PopperProps

export function ContextMenu({
  children,
  menuItems,
  menuStyleProps,
  onLeftClick = false,
  alignContentLeft = false,
  disabled = false,
  ...rest
}: PropsWithChildren<ContextMenuProps>): JSX.Element {
  const containerRef = useRef<HTMLDivElement>(null)
  const triggerContainerRef = useRef<HTMLDivElement>(null)
  const [showMenu, setShowMenu] = useState(false)
  const [menuPosition, setMenuPosition] = useState({ x: 0, y: 0 })

  useOnClickOutside(containerRef, () => {
    setShowMenu(false)
  })

  const onContextMenu = (e: React.MouseEvent<HTMLDivElement>): void => {
    if (disabled) {
      return
    }

    e.preventDefault()
    e.stopPropagation()
    setShowMenu(true)

    // Capture raw click coords
    const { clientX, clientY } = e
    setMenuPosition({ x: clientX, y: clientY })
  }

  function getRelativeCoordinates(): { x: number; y: number } {
    if (!triggerContainerRef.current) {
      return { x: 0, y: 0 }
    }

    const rect = triggerContainerRef.current.getBoundingClientRect()
    const relativeX = alignContentLeft ? menuPosition.x - rect.right : menuPosition.x - rect.left
    const relativeY = menuPosition.y - rect.top - rect.height

    return {
      x: relativeX,
      y: relativeY,
    }
  }

  const { x, y } = getRelativeCoordinates()

  if (disabled) {
    return <Fragment>{children}</Fragment>
  }

  return (
    <Popover
      allowFlip
      open={showMenu}
      strategy="absolute"
      placement={alignContentLeft ? 'bottom-end' : 'bottom-start'}
      offset={{ mainAxis: y, crossAxis: x }}
      {...rest}
    >
      {/*
        We attach the context menu event handler conditionally:
        - If onLeftClick is true, we use onMouseDown to open the menu on left-click.
        - Otherwise, we use onContextMenu to open the menu on right-click.
        This ensures that left-click events are not blocked from propagating,
        keeping normal click behavior intact.
      */}
      <Popover.Trigger onMouseDown={onLeftClick ? onContextMenu : undefined}>
        <div ref={triggerContainerRef} onClick={undefined} onContextMenu={onLeftClick ? undefined : onContextMenu}>
          {children}
        </div>
      </Popover.Trigger>
      <Popover.Content
        ref={containerRef}
        key={`${menuPosition.x}-${menuPosition.y}`} // This key ensures that the component re-renders when the menu position changes so we get a re-animation
        backgroundColor="transparent"
        enterStyle={{
          opacity: 0,
          scale: 0.98,
          transform: [{ translateY: -4 }],
          animation: '125ms',
        }}
      >
        <MenuContent
          items={menuItems}
          onItemClick={() => {
            setShowMenu(false)
          }}
          {...menuStyleProps}
        />
      </Popover.Content>
    </Popover>
  )
}
