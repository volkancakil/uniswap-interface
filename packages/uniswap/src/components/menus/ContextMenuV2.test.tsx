import { act, fireEvent } from '@testing-library/react-native'
// eslint-disable-next-line no-restricted-imports
import { View } from 'react-native'
import { ContextMenu, MenuOptionItem } from 'uniswap/src/components/menus/ContextMenuV2'
import { renderWithProviders } from 'uniswap/src/test/render'

describe('ContextMenu', () => {
  const mockMenuItems: MenuOptionItem[] = [
    {
      label: 'Option 1',
      onPress: jest.fn(),
    },
    {
      label: 'Option 2',
      onPress: jest.fn(),
      showDivider: true,
    },
  ]

  it('renders without error', () => {
    const tree = renderWithProviders(
      <ContextMenu menuItems={mockMenuItems}>
        <div>Trigger</div>
      </ContextMenu>,
    )

    expect(tree).toMatchSnapshot()
  })

  describe('opens the menu', () => {
    it('on right-click by default', () => {
      const { getByTestId, queryByText } = renderWithProviders(
        <ContextMenu menuItems={mockMenuItems}>
          <View testID="trigger">
            <div>Trigger</div>
          </View>
        </ContextMenu>,
      )

      const trigger = getByTestId('trigger')

      act(() => {
        fireEvent(trigger, 'contextMenu', {
          preventDefault: jest.fn(),
          stopPropagation: jest.fn(),
          clientX: 100,
          clientY: 150,
        })
      })

      expect(queryByText('Option 1')).toBeTruthy()
      expect(queryByText('Option 2')).toBeTruthy()
    })

    it('on left-click if onLeftClick is true', () => {
      const { getByTestId, queryByText } = renderWithProviders(
        <ContextMenu menuItems={mockMenuItems} onLeftClick>
          <View testID="trigger">
            <div>Trigger</div>
          </View>
        </ContextMenu>,
      )

      const trigger = getByTestId('trigger')

      act(() => {
        fireEvent(trigger, 'onMouseDown', {
          preventDefault: jest.fn(),
          stopPropagation: jest.fn(),
          clientX: 100,
          clientY: 150,
        })
      })
      expect(queryByText('Option 1')).toBeTruthy()
      expect(queryByText('Option 2')).toBeTruthy()
    })
  })

  describe('handles edge cases', () => {
    it('does not open the menu if no menuItems are provided', () => {
      const { getByTestId, queryByRole } = renderWithProviders(
        <ContextMenu menuItems={[]}>
          <View testID="trigger">
            <div>Trigger</div>
          </View>
        </ContextMenu>,
      )

      const trigger = getByTestId('trigger')

      act(() => {
        fireEvent.press(trigger)
      })

      const menu = queryByRole('menu')
      expect(menu).toBeFalsy()
    })

    it('does not crash if trigger element is not found', () => {
      const { queryByRole } = renderWithProviders(
        <ContextMenu menuItems={mockMenuItems}>
          <View testID="trigger">
            <div>Trigger</div>
          </View>
        </ContextMenu>,
      )

      const menu = queryByRole('menu')
      expect(menu).toBeFalsy()
    })
  })
})
