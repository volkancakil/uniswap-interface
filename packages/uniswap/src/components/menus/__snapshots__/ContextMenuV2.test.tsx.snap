// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu renders without error 1`] = `
<View
  aria-expanded={false}
  cancelable={true}
  disabled={false}
  focusable={true}
  minPressDuration={0}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
>
  <div
    onContextMenu={[Function]}
  >
    <div>
      Trigger
    </div>
  </div>
</View>
`;
